### PR TITLE
add wikitext2_gpt2 benchmark pipeline

### PR DESCRIPTION
**Pull request by haochend413**
### 1. Motivation and Context

Add wikitext_gpt2 experiment benchmark pretrained models & data pipeline for 0.3.0 release. 

### 2. Summary of the change

Added wikitext_gpt2 option to load_benchmark. Same model, train and loss function implementations as in gpt2_wikitext experiment. 

### 3. What tests have been added/updated for the change?
- [ ] N/A: No test will be added (please justify)
- [ ] Unit test: Typically, this should be included if you implemented a new function/fixed a bug.
- [ ] Application test: If you wrote an example for the toolkit, this test should be added.
- [ ] Document test: If you added an external API, then you should check if the document is correctly generated.
- [ ] ...

### CODE REVIEW

File: dattri/benchmark/datasets/wikitext/data.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+from __future__ import annotations
+from typing import Tuple
+import torch
+from datasets import load_dataset
+from transformers import AutoTokenizer
```
Comment (TheaperDeng) [inline dattri/benchmark/datasets/wikitext/data.py]: Add `transformers` and `datasets` to https://github.com/TRAIS-Lab/dattri/blob/main/pyproject.toml#L33

File: dattri/benchmark/load.py [hunk 2]
```diff
@@ -97,6 +110,7 @@ def generate_url_map(identifier: str) -> Dict[str, Any]:
     "cifar2_resnet9": create_resnet9_model,
     "shakespeare_nanogpt": None,
     "maestro_musictransformer": create_musictransformer_model,
+    "wikitext_gpt2": create_gpt2_model,
 }
```
Comment (TheaperDeng) [inline dattri/benchmark/load.py]: change the load api as well. including the sampler.

File: examples/pretrained_benchmark/logra_wikitext_gpt2_lds.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    # download the pre-trained benchmark
+    # includes some trained model and ground truth
+    model_details, groundtruth = load_benchmark(
+        model="gpt2", dataset="wikitext2", metric="lds"
+    )
+
+    def f(model, batch, device):
+        # inputs, targets = batch
+        # inputs = inputs.to(device)
+        # targets = targets.to(device)
```
Comment (TheaperDeng) [inline examples/pretrained_benchmark/logra_wikitext_gpt2_lds.py]: delete these comments

File: examples/pretrained_benchmark/logra_wikitext_gpt2_lds.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    score = attributor.attribute(
+        DataLoader(
+            model_details["train_dataset"],
+            batch_size=4681,
+            sampler=model_details["train_sampler"],
+        ),
+        DataLoader(
+            model_details["test_dataset"],
+            batch_size=4681,
```
Comment (TheaperDeng) [inline examples/pretrained_benchmark/logra_wikitext_gpt2_lds.py]: test does not have 4681 samples

File: examples/pretrained_benchmark/logra_wikitext2_gpt2_lds.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            config=AutoConfig.from_pretrained("openai-community/gpt2"),
+            from_tf=False,
+        )
+
+        # load from stored pretrained.
+        state_dict = torch.load(checkpoint, map_location="cpu")
+        model.load_state_dict(state_dict, strict=True)
+        model = replace_conv1d_modules(model)
+
+        model.cuda()
```
Comment (TheaperDeng) [inline examples/pretrained_benchmark/logra_wikitext2_gpt2_lds.py]: Then we can not use other device right?

File: dattri/benchmark/load.py [hunk 6]
```diff
@@ -266,9 +285,13 @@ def load_benchmark(  # noqa:PLR0914
...
     if SUPPORTED_DATASETS[dataset] is not None:
-        train_dataset, test_dataset = SUPPORTED_DATASETS[dataset](
-            download_path / "dataset",
-        )
+        if dataset == "wikitext2":
+            train_dataset, test_dataset = SUPPORTED_DATASETS[dataset]()
+        else:
+            train_dataset, test_dataset = SUPPORTED_DATASETS[dataset](
+                download_path / "dataset",
+            )
```
[the remark this answers is not in the document]
Comment (haochend413) [reply dattri/benchmark/load.py]: Added a path argument to specify cache directory for wikitext dataset and removed the branch.

File: dattri/algorithm/block_projected_if/block_projected_if.py [hunk 7]
```diff
@@ -247,7 +247,12 @@ def cache(self, full_train_dataloader: DataLoader) -> None:
             loss = self.task.original_loss_func(self.model, batch, self.device)
 
             # Get actual batch size (the last batch may have fewer samples)
-            batch_size = batch[0].shape[0]
+            if isinstance(batch, dict):
+                # handle dataloader that returns dict
+                # mainly to support default_data_collator in transformers
+                batch_size = next(iter(batch.values())).shape[0]
+            else:
+                batch_size = batch[0].shape[0]
```
[the remark this answers is not in the document]
Comment (TheaperDeng) [reply dattri/algorithm/block_projected_if/block_projected_if.py]: Yeah, let me revise this a bit. Though it can not be super robust since the Mapping type may have non-tensor value
